app/src/main/java/org/sil/gatherwords/SessionActivity.java

```diff
@@ -118,18 +130,121 @@ public void save_settings_fab_pressed(View view) {
+        }
+
+        @Override
+        protected List<Long> doInBackground(Void... voids) {
+            List<Long> ids = new ArrayList<>();
+            try {
+                // Wait for the insert session task to finish and get the result (the id(s) of the session)
+                ids = task.get();
+            } catch (InterruptedException | ExecutionException e) {
+                Log.getStackTraceString(e);
+            }
+            if (ids.size() == 1) {
+                sessionID = ids.get(0);
+            }
+            insertFromAsset();
+
             return null;
         }
+
+        @Override
+        protected void onProgressUpdate(Double... values) {
+            super.onProgressUpdate(values);
+            double progress = (values[0] / maxProgress)*100;
+            CharSequence sequence = new DecimalFormat("##.##%").format(progress);
+            // Too slow!!! TODO get quick progress updates
+//            Toast.makeText(sessionActivity.getApplicationContext(), sequence, Toast.LENGTH_SHORT).show();
+        }
+
+        /**
+         * Insert the words from the selected asset or exit if none selected
+         */
+        private void insertFromAsset() {
+            if (sessionActivity.get().worldListToLoad.equals("")) {
+                // There was no file selected
+                return;
+            }
+
+            // Begin the transaction
+            sessionActivity.get().db.beginTransaction();
+            try {
+                JSONArray jsonArray = new JSONArray(loadWordList());
+                maxProgress = jsonArray.length();
+
+                // Iterate through each word from the file
+                for (int i = 0; i < jsonArray.length(); i++) {
+                    JSONObject json = jsonArray.getJSONObject(i);
+                    Word word = new Word();
+
+                    // Link the word to the session that was just created
+                    word.sessionId = sessionID;
+                    word.meanings = json.getString("entry");
+
+                    // Insert
+                    wordDao.insertWords(word);
+                    this.publishProgress((double) i);
+                }
+
+                // Mark to commit the changes to the DB
+                sessionActivity.get().db.setTransactionSuccessful();
+            } catch (Exception e) {
+                Log.getStackTraceString(e);
+            }
+            finally {
+                // Commit or rollback the database
+                sessionActivity.get().db.endTransaction();
+            }
+        }
+
+        /**
+         * Read the selected file
+         *
+         * @return The contents of the file
+         */
+        private String loadWordList() {
+            String file = "";
+            try {
+                InputStream is = sessionActivity.get().getApplicationContext().getAssets().open("wordLists/" + sessionActivity.get().worldListToLoad);
+                byte[] buffer = new byte[is.available()];
```

Comment:
This is not guaranteed to be the entire file.  Probably safer to loop with a `StringBuilder`.